src/dotnet/UI.Blazor/Services/HistoryUI/HistoryUI.cs

```diff
@@ -0,0 +1,199 @@
+        // We can get rid of this event handler
+        // after OnLocationChangedAsync is replaced with synchronous implementation in .NET 7.
+        => await _whenLocationChangedHandled;
+
+    public Task GoBack()
+        => JS.InvokeVoidAsync("eval", "history.back()").AsTask();
+
+    public void NavigateTo(Action? onForwardAction, Action? onBackAction)
+    {
+        if (_pendingHistoryItem != null)
+            throw StandardError.Constraint("There is still pending history item");
+        var historyItem = new HistoryItem(Nav.Uri) {
+            OnForwardAction = onForwardAction,
+            OnBackAction = onBackAction
+        };
+        _pendingHistoryItem = new PendingHistoryItem(historyItem, _historyIndex);
+        Nav.NavigateTo(Nav.Uri);
+    }
+
+    private async Task<State?> GetStateAsync()
+    {
+        await _whenJsRefInitialized.ConfigureAwait(false);
+        var result = await _jsRef!.InvokeAsync<State?>("getState").ConfigureAwait(false);
+        return result;
+    }
+
+    private async Task InitializeJsRef()
+        => _jsRef = await JS.InvokeAsync<IJSObjectReference>(
+            $"{BlazorUICoreModule.ImportName}.HistoryUI.create")
+            .ConfigureAwait(false);
+
+    private async Task InitializeState(State state)
+    {
+        await _whenJsRefInitialized;
+        await SetStateAsync(state);
+    }
+
+    private void OnLocationChanged(object? sender, LocationChangedEventArgs e)
+    {
+        IsInitialLocation = false;
+
+        var location = e.Location;
+        Log.LogDebug("Location changed: '{Location}'", location);
+
+        if (!BrowserInfo.ScreenSize.Value.IsNarrow())
+            return;
+
+        var tcs = new TaskCompletionSource();
+        _whenLocationChangedHandled = tcs.Task;
+
+        _ = OnLocationChangedAsync(tcs, e.Location);
+    }
+
+    private async Task OnLocationChangedAsync(TaskCompletionSource whenCompleted, string location)
+    {
+        try {
+            var previousState = _state;
+            HistoryMove move;
+            // TODO: in .NET 7 NavigationManager provides access to state property of the history API.
+            // Rework history position tracking with it.
+            var readState = await GetStateAsync();
+            if (readState == null) {
+                // State is null, apparently it's an internal navigation to next location happened.
+                // Lets store this in the state.
+                var state = new State { Index = previousState.Index + 1 };
+                await SetStateAsync(state);
+                _state = state;
+                move = HistoryMove.Navigate;
+            }
+            else {
+                move = readState.Index < previousState.Index
+                    ? HistoryMove.Backward
+                    : readState.Index > previousState.Index
+                        ? HistoryMove.Forward
+                        : throw StandardError.Constraint("Invalid state index for history move.");
```

Comment:
This error pops up in dev when you press "show left panel" button in narrow mode.

Reply:
Hm. I can not reproduce.
Can you describe repro steps?

Reply:
Yes:
- Switch to narrow mode in Chrome
- Click N times on "<-" button at the top-left corner.

Reply:
Still can not reproduce. May be it's connected with races.
Do you switch to narrow mode in Chrome on blank page and then open https://dev.actual.chat/ ?
After that you navigates to chats and back?